config/crds/autoscaling_v1alpha1_hvpa.yaml

```diff
@@ -54,6 +54,16 @@ spec:
                     to which the autoscaler can scale down. It defaults to 1 pod.
                   format: int32
                   type: integer
+                updatePolicy:
```

Comment:
Did you get a change to take a look at the proposed structure for the spec [here](https://github.com/gardener/hvpa-controller/issues/13#issuecomment-518613615)?

Reply:
Inspired from [VPA](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2/types.go#L91).
IMHO, we can keep it same, unless there is a strong reason to change it. WDYT?

Reply:
The question was not about the `PodUpdatePolicy` structure with the embedded `updateMode` structure as in VPA. That is fine.

My question was more about the other stuff about the template including metadata, selector etc. Inspired by [`ReplicaSetSpec`](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/apps/types.go#L722).

Reply:
In that case how about this:
```
apiVersion: ...
kind: HVPA
metadata: {}
spec:
  selector:
  targetRef:
  scaleUpStabilizationWindow:
  scaleDownStabilizationWindow:
  WeightBasedScalingIntervals:
  hpa:
    updatePolicy: Off/Auto/ScaleOut
    template:
      metadata: {...labels...}
      spec:
        ...
  vpa:
    updatePolicy: Off/Auto/ScaleUp
    template:
      metadata: {...labels...}
      spec:        
        resourcePolicy:
          containerPolicies:
          - containerName: ...
             mode: On/Auto
```

Reply:
If metadata is specifiable separately, we have to assume the worst that the user might specify the labels differently for VPA and HPA. Then a common `selector` would not work.

So, either it is common `selector` and `metadata`/`labels` or it is separate `selector` and `metadata`/`labels`. It cannot be any other combination.

How would the spec look like then? Like below?

```yaml
apiVersion: ...
kind: HVPA
metadata: {}
spec:
  selector:
  targetRef:
  scaleUpStabilizationWindow:
  scaleDownStabilizationWindow:
  WeightBasedScalingIntervals:
  template:
    metadata: {...labels...}
  hpa:
    updatePolicy: Off/Auto/ScaleOut
    spec:
      ...
  vpa:
    updatePolicy: Off/Auto/ScaleUp
    spec:        
      resourcePolicy:
        containerPolicies:
        - containerName: ...
           mode: On/Auto
```

Looks very ugly like a [deconstructed samosa](https://www.sanjeevkapoor.com/recipe/Deconstructed-Samosa.html)😉

Reply:
I am not sure how the alternative is any different (+/- 1 or 2 fields):
```yaml
apiVersion: ...
kind: HVPA
metadata: {}
spec:
  targetRef:
  scaleUpStabilizationWindow:
  scaleDownStabilizationWindow:
  WeightBasedScalingIntervals:
  hpa:
    selector:
    updatePolicy: Off/Auto/ScaleOut
    template:
      metadata: {}
      spec:
        ...
  vpa:
    selector:
    updatePolicy: Off/Auto/ScaleUp
    template:
      metadata: {}
      spec:        
        resourcePolicy:
          containerPolicies:
          - containerName: ...
             mode: On/Auto
```
Regardless, IMHO, looking like a deconstructed samosa still does not warrant duplication of fields.

Reply:
In the [above structure](https://github.com/gardener/hvpa-controller/pull/19/files#r312841081), there is no correlation between what applies to what.
- `template` and `selector` apply to `hpa` and `vpa` as an aggregation (hpa and vpa as owned by hvpa), 
- The rest of the fields are somewhat unrelated in terms of aggregation but related in terms of logic.

Besides, that point if template is to give the user the flexibility about how to label these resources. There are no good reasons to say we want to force the user to use the same labels for both HPA and VPA other than convenience of implementation. And that would not be a good reason.

Also, this argument would hold even for the `ReplicaSetSpec` as to why the `metadata` for the template as well as the selector in the `ReplicaSetSpec` be possible to be configured in the first place? Why not simply inherit the `metadata` from the `ReplicaSet`'s own `metadata? The argument against this approach in `ReplicaSetSpec` would hold in `HVPASpec` equally.

To cut a long argument short, IMHO, it would be better to keep the `template` and `selector` separate for `hpa` and `vpa` in `HVPASpec`.

Reply:
Thanks Amshu for explaining. I find these reasons valid. These are exactly the reasons i kept `templates` separate [here](https://github.com/gardener/hvpa-controller/pull/19#discussion_r312814855).
I was just concerned about the unnecessarily separate `selector` fields. Will keep them separate as you suggested.